### PR TITLE
Fix spurious nil derefs on retrying local activities

### PR DESCRIPTION
**Pull request by Sushisource**
Depends on https://github.com/temporalio/sdk-go/pull/302 being merged first
Relates to https://www.notion.so/Go-SDK-Nil-Reference-Panic-in-Bench-Test-b240b5f0db3b4d8088e3486939a53e78

~It appears my fix in the MR this one depends on also fixes this item. I wasn't able to repro the issue after a variety of bench runs.~ Not quite. Will require a bit of extra fix.

### CODE REVIEW

File: internal/internal_task_handlers.go [hunk 1]
```diff
@@ -494,15 +493,10 @@ func (w *workflowExecutionContextImpl) Unlock(err error) {
...
-	if eventHandler == nil {
+	if w.eventHandler == nil {
 		return nil
 	}
-	eventHandlerImpl, ok := eventHandler.(*workflowExecutionEventHandlerImpl)
-	if !ok {
-		panic("unknown type for workflow execution event handler")
-	}
-	return eventHandlerImpl
+	return (*w.eventHandler).(*workflowExecutionEventHandlerImpl)
```
Comment (wxing1292) [inline internal/internal_task_handlers.go:499]: do we need to grab a lock for this?

Comment (wxing1292) [reply internal/internal_task_handlers.go:499]: since previously this is stored in atomic.Value

Comment (Sushisource) [reply internal/internal_task_handlers.go:499]: Putting it here can cause deadlocks - generally functions calling into this one are doing the locking. Overall there's not much consistency around when the lock is taken. I'd be worth cleaning this up as part of the bigger refactorings I'll take on soon.

File: internal/internal_task_handlers_test.go [hunk 2]
```diff
@@ -1214,6 +1215,97 @@ func (t *TaskHandlersTestSuite) TestWorkflowTask_PageToken() {
 	t.NotNil(response)
 }
 
+func (t *TaskHandlersTestSuite) TestLocalActivityRetry_Workflow() {
```
Comment (vitarb) [inline internal/internal_task_handlers_test.go:1218]: Should we add a test that validates scenario mentioned in https://github.com/uber-go/cadence-client/issues/915?

Comment (Sushisource) [reply internal/internal_task_handlers_test.go:1218]: This one doesn't - but the existing one directly after it does, so I think the path is covered.

Comment (vitarb) [reply internal/internal_task_handlers_test.go:1218]: Yeah, I see, there is `TestLocalActivityRetry_WorkflowTaskHeartbeatFail`. Looks good to me then.

**Review by vitarb — APPROVED** [review]

**Pull request merged**